### PR TITLE
[web_editor] : Migrate to new api

### PR DESCRIPTION
**Pull request by mba-odoo**
Task: https://www.odoo.com/web#id=21484&view_type=form&model=project.task&action=327&menu_id=4720
Pad: https://pad.odoo.com/p/r.fBnXPC7m7XeB6b8n

### CODE REVIEW

File: addons/web_editor/controllers/main.py [hunk 1]
```diff
@@ -135,23 +135,23 @@ def attach(self, func, upload=None, url=None, disable_optimization=None, **kwarg
...
                 'public': True,
                 'res_model': 'ir.ui.view',
-            }, request.context)
-            uploads += Attachments.read(request.cr, request.uid, [attachment_id], ['name', 'mimetype', 'checksum', 'url'], request.context)
+            })
+            uploads += attachment.read(['name', 'mimetype', 'checksum', 'url'])
         else:                                                  # images provided
             try:
-                attachment_ids = []
+                attachments = Attachments.browse([])
```
Comment (msh-odoo) [inline addons/web_editor/controllers/main.py]: do we need to browse, ``` attachments = request.env['ir.attachment'] ``` do the same job

File: addons/web_editor/models/ir_qweb.py [hunk 2]
```diff
@@ -3,39 +3,35 @@
...
 """
-
 import cStringIO
-import datetime
+import hashlib
 import itertools
 import json
 import logging
 import os
+import pytz
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: Put pytz in third party group

File: addons/web_editor/models/ir_qweb.py [hunk 3]
```diff
@@ -3,39 +3,35 @@
...
-from openerp.osv import orm, fields
-from openerp.tools import ustr, DEFAULT_SERVER_DATE_FORMAT, DEFAULT_SERVER_DATETIME_FORMAT
-from openerp.tools import html_escape as escape
-from openerp.addons.base.ir import ir_qweb
-from openerp.tools.translate import translate
+import odoo
+from odoo import api, fields, models
+from odoo.addons.base.ir import ir_qweb
+from odoo.tools import html_escape as escape
+from odoo.tools import ustr
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: Merge this two imports

File: addons/web_editor/models/ir_qweb.py [hunk 4]
```diff
@@ -44,11 +40,11 @@ class QWeb(orm.AbstractModel):
     re_remove_spaces = re.compile('\s+')
 
     def render_tag_snippet(self, element, template_attributes, generated_attributes, qwebcontext):
-        cr = qwebcontext['request'].cr
-        uid = qwebcontext['request'].uid
-        view_reg = self.pool['ir.ui.view']
+        # method does not have cr, uid, and context as argument so can not apply api.models and other decorator so to access env just create env
+        self.env = api.Environment(qwebcontext['request'].cr, qwebcontext['request'].uid, qwebcontext['request'].context)
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: Do not override environment of self

File: addons/web_editor/controllers/main.py [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
-from openerp import tools
 import cStringIO
-import werkzeug.wrappers
-import time
+import io
+import json
 import logging
-logger = logging.getLogger(__name__)
+import time
+import werkzeug.wrappers
```
Comment (msh-odoo) [inline addons/web_editor/controllers/main.py]: separate standard imports and third party imports

File: addons/web_editor/controllers/main.py [hunk 6]
```diff
@@ -39,17 +38,18 @@ def FieldTextHtml(self, model=None, res_id=None, field=None, callback=None, **kw
...
 
         trans = dict(
-            lang=kwargs.get('lang', context.get('lang')),
+            lang=kwargs.get('lang', request.env.context.get('lang')),
             translatable=kwargs.get('translatable'),
             edit_translations=kwargs.get('edit_translations'),
             editable=kwargs.get('enable_editor'))
 
-        context.update(trans)
+        local_context = dict(request.env.context or {})
```
Comment (msh-odoo) [inline addons/web_editor/controllers/main.py]: You can omit or part

Comment (sindhav) [inline addons/web_editor/controllers/main.py]: @msh-odoo  need ```or {}``` otherwise raise error if request.env.context is False

Comment (msh-odoo) [inline addons/web_editor/controllers/main.py]: @sindhav Its never gonna be False

File: addons/web_editor/models/ir_qweb.py [hunk 7]
```diff
@@ -57,6 +53,14 @@ def render_tag_snippet(self, element, template_attributes, generated_attributes,
         template_attributes['call'] = template_attributes['snippet']
         return self.render_tag_call(element, template_attributes, generated_attributes, qwebcontext)
 
+    @api.v8
+    def get_converter_for(self, field_type):
+        if self.env.__contains__('ir.qweb.field.' + field_type):
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: Do it this way:
``` if 'ir.qweb.field.' + field_type in self.env:
            return self.env['ir.qweb.field.' + field_type]
        return self.env['ir.qweb.field'] ```
Do not call __contains__ method, you can use in operator which will do the same

File: addons/web_editor/models/ir_qweb.py [hunk 8]
```diff
@@ -83,42 +88,41 @@ def attributes(self, cr, uid, field_name, record, options,
         if placeholder:
             attrs.append(('placeholder', placeholder))
 
-        if context and context.get('edit_translations') and context.get('translatable') and field.type in ('char', 'text') and field.translate:
+        if self.env.context and self.env.context.get('edit_translations') and self.env.context.get('translatable') and field.type in ('char', 'text') and field.translate:
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: You can directly write:
``` if self.env.context.get('edit_translations') and self.env.context.get('translatable') and field.type in ('char', 'text') and field.translate: ```

File: addons/web_editor/models/ir_translation.py [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
 
-from openerp import models, api
+from odoo import models, api
 
-class ir_translation(models.Model):
+class IrTranslation(models.Model):
     _inherit = 'ir.translation'
 
     @api.model
     def _get_terms_mapping(self, field, records):
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_translation.py]: Do we need decoration now ?

Comment (mba-odoo) [inline addons/web_editor/models/ir_translation.py]: @msh-odoo  no base method is also in new api so not required decorator. thanks

File: addons/web_editor/models/ir_qweb.py [hunk 10]
```diff
@@ -144,43 +145,43 @@ def attributes(self, cr, uid, field_name, record, options,
...
 
     # helper to call the rendering of contact field
-    def get_record_to_html(self, cr, uid, ids, options=None, context=None):
-        node = self.record_to_html(cr, uid, 'record', {
-            'record': self.pool['res.partner'].browse(cr, uid, ids[0], context=context)},
-            options=options, context=context)
+    @api.model
+    def get_record_to_html(self, ids, options=None):
+        node = self.record_to_html('record', {
+            'record': self.env['res.partner'].browse(ids[0])},
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: I think this method is only called from JS with single ID why not making this method handling single ID(as it is used for m2o)

File: addons/web_editor/models/ir_qweb.py [hunk 11]
```diff
@@ -195,54 +196,48 @@ def parse_fuzzy(in_format, value):
...
         if isinstance(value, basestring):
-            value = datetime.datetime.strptime(
-                value, DEFAULT_SERVER_DATETIME_FORMAT)
+            value = fields.Datetime.from_string(value)
         if value:
             # convert from UTC (server timezone) to user timezone
-            value = fields.datetime.context_timestamp(
-                cr, uid, timestamp=value, context=context)
-            value = value.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
+            value = fields.Datetime.context_timestamp(self, timestamp=value)
```
Comment (msh-odoo) [inline addons/web_editor/models/ir_qweb.py]: Don't you think we should pass record instead of self  in context_timestamp ?

**Comment by hmo-odoo** [comment]
continue work on this PR: https://github.com/odoo/odoo/pull/11682

**Pull request closed without merging**